From f8cb397ee68cc274262c09a2dfc0ad5fb02983c9 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:11:16 -0400
Subject: [powerpc] Fix Oops in rtas_stop_self()

Message-id: <1410545655-205645-448-git-send-email-dzickus@redhat.com>
Patchwork-id: 94648
O-Subject: [RHEL7 PATCH 447/626] powerpc: Fix Oops in rtas_stop_self()
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 4fb8d027dca0236c811272d342cf185569d91311
Author: Li Zhong <zhong@linux.vnet.ibm.com>
Date:   Mon Apr 28 08:29:51 2014 +0800

    powerpc: Fix Oops in rtas_stop_self()

    commit 41dd03a9 may cause Oops in rtas_stop_self().

    The reason is that the rtas_args was moved into stack space. For a box
    with more that 4GB RAM, the stack could easily be outside 32bit range,
    but RTAS is 32bit.

    So the patch moves rtas_args away from stack by adding static before
    it.

    Signed-off-by: Li Zhong <zhong@linux.vnet.ibm.com>
    Signed-off-by: Anton Blanchard <anton@samba.org>
    Cc: stable@vger.kernel.org # 3.14+
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/platforms/pseries/hotplug-cpu.c b/arch/powerpc/platforms/pseries/hotplug-cpu.c
index 9b8e050..20d6297 100644
--- a/arch/powerpc/platforms/pseries/hotplug-cpu.c
+++ b/arch/powerpc/platforms/pseries/hotplug-cpu.c
@@ -88,13 +88,14 @@ void set_default_offline_state(int cpu)
 
 static void rtas_stop_self(void)
 {
- struct rtas_args args = {
-  .token = cpu_to_be32(rtas_stop_self_token),
+ static struct rtas_args args = {
   .nargs = 0,
   .nret = 1,
   .rets = &args.args[0],
  };
 
+ args.token = cpu_to_be32(rtas_stop_self_token);
+
  local_irq_disable();
 
  BUG_ON(rtas_stop_self_token == RTAS_UNKNOWN_SERVICE);
-- 
1.7.1